Remove problematic sort in m2_to_m2

adrianeboyd opened the pull request:
Remove sort that reorders multiple insertions at the same token position. Preserve original order instead of reordering tokens alphabetically.

Commit:
Remove problematic sort in m2_to_m2

Remove sort that reorders multiple insertions at the same token
position.

chrisjbryant commented:
I noticed this recently too. Does everything still work as intended without the sorting? I guess I put it there originally cause I was worried the edits weren't an OrderedDict.

adrianeboyd commented:
Okay, now I see that I wasn't right about which sort was causing problems. It's actually this one:

https://github.com/chrisjbryant/errant/blob/1ba32f9fc3f5e7dd8937f5a702b8f96e53978383/scripts/toolbox.py#L46

(I wasn't actually using m2_to_m2 itself, but I adapted this loop from m2_to_m2 to create a m2_to_parallel script to convert m2 with one annotator back to a parallel corpus. There I definitely had multiple inserts reordered alphabetically and I thought removing this sort fixed the problem, but when I looked closer I was mistaken.)

Here's a test case to show the problems (`really` has a weird category on purpose):

```
S This are a great sentences .
A 1 1|||R:VERB|||really|||REQUIRED|||-NONE-|||0
A 1 1|||R:ADV|||truly|||REQUIRED|||-NONE-|||0
A 1 1|||R:ADV|||definitely|||REQUIRED|||-NONE-|||0
A 1 2|||R:VERB:SVA|||is|||REQUIRED|||-NONE-|||0
A 4 5|||R:NOUN:NUM|||sentence|||REQUIRED|||-NONE-|||0
A 0 1|||R:DET|||These|||REQUIRED|||-NONE-|||1
A 2 3|||U:DET||||||REQUIRED|||-NONE-|||1
```

The current `-gold -old_cats` m2_to_m2 output:

```
S This are a great sentences .
A 1 1|||R:ADV|||definitely|||REQUIRED|||-NONE-|||0
A 1 1|||R:ADV|||truly|||REQUIRED|||-NONE-|||0
A 1 1|||R:VERB|||really|||REQUIRED|||-NONE-|||0
A 1 2|||R:VERB:SVA|||is|||REQUIRED|||-NONE-|||0
A 4 5|||R:NOUN:NUM|||sentence|||REQUIRED|||-NONE-|||0
A 0 1|||R:DET|||These|||REQUIRED|||-NONE-|||1
A 2 3|||U:DET||||||REQUIRED|||-NONE-|||1
```

I can't find any supporting M2 documentation for these kinds of inserts, but it does seem like you have to allow for multiple insertions at the same position because there could be multiple words related to separate errors at the same position. (Examples with an inserted word followed by an inserted comma are common and clearly have separate error types.)

Without any other ordering information in the M2 format, I think this means that you have to require that the M2 input has the edits for a single coder in order. I suppose you could potentially preserve the order within a sublist for each start position (per coder) and just sort on a list of per-start-position lists instead, but that's getting kind of overly complicated. (Do you have a lot of M2 reference corpora with unordered edits? How did this error did this come up? Hand edits to existing M2 data?)

Maybe you could check that sorting a list of A start/end position pairs for a single coder doesn't change anything (about the pairs themselves, so just sorting `[[1, 1], [1, 1], [1, 2], [4, 5]]` etc.)  and throw an exception otherwise?

The processM2() function keeps lists (not dicts) of errors for each coder in the order they were given the input, so if you're requiring edits to be in order in the input I don't think you need to change much here other than removing the sort.

In general, you're right that you'd shouldn't be relying on the fact that an iterator currently iterates over a dictionary in the order that the entries were added, but you need to make sure there's a good key to sort on in the first place. The M2 format definitely doesn't provide many good keys for sorting, but I think the sort I was initially worried about here just sorts by coder ID, which seems correct for m2_to_m2.

Commit:
Remove correct problematic sort

Update to reinsert previously removed sort in m2_to_m2 and remove the
sort from toolbox.processM2() that's actually causing the token
reordering problem.

chrisjbryant commented:
Great, thanks for the detailed breakdown! I knew there was a reason I had the sort in there originally, but you helped remind me!

I think the standard `sorted` function sorts based on every item in the list, so if the start and end token boundaries are the same, it ends up sorting alphabetically by error type instead. 
I've just pushed an update to `processM2` however which means it'll now just sort by the start and end spans alone. Since python sorts are stable, this means items with the same start and end span are never reordered. I did some tests to make sure this was the case, but let me know if you come across any problems.

Thanks for the help!
